File: LeetCode/0034-Find first & last pos of element in a sorted arr/main.cpp
//
// Created by 薛智钧 on 2020/3/17.
//

#include <iostream>
#include <vector>
using namespace std;

// Given an array of integers nums sorted in ascending order,
// find the starting and ending position of a given target value.
// Your algorithm's runtime complexity must be in the order of O(log n).
// If the target is not found in the array, return [-1, -1].

class Solution {
public:
    // 并不是严格的o(logn)
    vector<int> searchRange(vector<int>& nums, int target) {
        if (nums.empty()) return {-1, -1};
        int end = binSearch_max(nums, target, 0, nums.size());
        if (end == -1) return {-1, -1};
        int start = end;
        for (; -1 < start; --start) {
            if (nums[start] < target)
                break;
        }
        return {start + 1, end};
    }

    vector<int> searchRange_2(vector<int>& nums, int target){
        if (nums.empty()) return {-1, -1};
        int end = binSearch_max(nums, target, 0, nums.size());
        if (end == -1) return {-1 , -1};
        int start = binSearch_min(nums, target, 0, nums.size());
        return {start, end};
    }

    // 返回命中元素最大的秩；若找不到，则返回-1
    int binSearch_max(vector<int>& nums, int target, int lo, int hi){
        while (1 < hi - lo){
            int mi = lo + (hi - lo) / 2; // 更安全，不会越界
            target < nums[mi] ? hi = mi : lo = mi; //[l0, mi) or [mi, hi)
        }
        return nums[lo] == target ? lo : -1;
    }

    // 返回命中元素最小的秩；若找不到，则返回-1
    int binSearch_min(vector<int>& nums, int target, int lo, int hi){
        while (lo < hi){
            int mi = lo + (hi - lo) / 2; // 更安全，不会越界
            target <= nums[mi] ? hi = mi : lo = mi + 1;
        }
        return nums[lo] == target ? lo : -1;
    }
};

int main(){
    vector<int> vec = {5,7,7,8,8,10};
    int target = 4;
    Solution sol;
    vector<int> res = sol.searchRange_2(vec, target);
    cout << "The begin and ending positions of target in the array are: ";
    for (int pos:res)
        cout << pos << ",";
}

File: LeetCode/0034-Find first & last pos of element in a sorted arr/main.py
"""  
Given an array of integers nums sorted in ascending order, 
find the starting and ending position of a given target value.

If target is not found in the array, return [-1, -1].

Follow up: Could you write an algorithm with O(log n) runtime complexity?


Example 1:

Input: nums = [5,7,7,8,8,10], target = 8
Output: [3,4]

"""


class Solution:
    # binary search with <=
    def searchRange1(self, nums: List[int], target: int) -> List[int]:
        if not nums:
            return [-1, -1]

        def searchLow():
            lo, hi = 0, len(nums) - 1
            while lo <= hi:
                mid = (lo+hi)//2
                if target <= nums[mid]:
                    hi = mid - 1
                else:
                    lo = mid + 1

            return lo

        def searchHigh():
            lo, hi = 0, len(nums) - 1
            while lo <= hi:
                mid = (lo+hi)//2
                if target < nums[mid]:
                    hi = mid - 1
                else:
                    lo = mid + 1

            return lo - 1

        min_rank = searchLow()
        max_rank = searchHigh()
        if 0 <= min_rank < len(nums) and min_rank <= max_rank and nums[min_rank] == target:
            return [min_rank, max_rank]

        return [-1, -1]

        ################################################################

    # binary search with <
    def searchRange2(self, nums: List[int], target: int) -> List[int]:
        if not nums:
            return [-1, -1]

        # find the lowest rank of the given target
        min_rank = self.binSearchFirst(nums, target)
        if min_rank == -1:
            return [-1, -1]

        # find the highest rank of the given target
        max_rank = self.binSearchLast(nums, target)

        return [min_rank, max_rank]

    def binSearchFirst(self, arr, target):
        lo, hi = 0, len(arr)
        while lo < hi:
            mid = (lo + hi)//2
            if target <= arr[mid]:
                hi = mid
            else:
                lo = mid + 1

        if lo == len(arr) or arr[lo] != target:
            return -1

        return lo

    def binSearchLast(self, arr, target):
        lo, hi = 0, len(arr)
        while lo < hi:
            mid = (lo + hi)//2
            if target < arr[mid]:
                hi = mid
            else:
                lo = mid + 1

        return lo - 1 if arr[lo-1] == target else -1
